src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+/// - `source`, `service`, `status`, `tags` and other reserved attributes are left as is;
+/// - the rest of the fields is moved to `attributes`.
+fn encode_event(event: Event) -> Option<EncodedEvent<PartitionInnerBuffer<Vec<u8>, Bytes>>> {
+    lazy_static! {
+        static ref KEY_PREFIX: Template =
+            Template::try_from("/dt=%Y%m%d/hour=%H/").expect("invalid object key format");
+    }
+    let key = KEY_PREFIX
+        .render_string(&event)
+        .map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key_prefix"),
+                drop_event: true,
+            });
+        })
+        .ok()?;
+
+    let mut log = event.into_log();
+
+    log.insert("_id", generate_log_id());
+
+    let timestamp = log
+        .remove(crate::config::log_schema().timestamp_key())
+        .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+    log.insert(
+        "date",
+        timestamp
+            .as_timestamp()
+            .cloned()
+            .unwrap_or_else(chrono::Utc::now)
+            .to_rfc3339_opts(SecondsFormat::Millis, true),
+    );
+    if let Some(message) = log.remove(crate::config::log_schema().message_key()) {
+        log.insert("message", message);
+    }
+    if let Some(message) = log.remove(crate::config::log_schema().host_key()) {
+        log.insert("host", message);
+    }
+
+    let mut attributes = BTreeMap::new();
+    let custom_attributes: Vec<String> = log
+        .keys()
+        .filter(|path| !RESERVED_ATTRIBUTES.contains(path.as_str()))
+        .collect();
+    for path in custom_attributes {
+        if let Some(value) = log.remove(&path) {
+            attributes.insert(path, value);
+        }
+    }
+    log.insert("attributes", attributes);
+
+    let bytes = serde_json::to_vec(&log)
+        .map(|mut b| {
+            b.push(b'\n');
+            b
+        })
+        .expect("Failed to encode event as json, this is a bug!");
+
+    Some(EncodedEvent {
+        item: PartitionInnerBuffer::new(bytes, key.into()),
+        finalizers: log.metadata_mut().take_finalizers(),
+    })
+}
+
+/// Generates a unique ID compatible with DD
+fn generate_log_id() -> String {
+    let mut id = BytesMut::with_capacity(18);
+    // timestamp in millis - 6 bytes
+    let now = Utc::now();
+    id.put_int(now.timestamp_millis(), 6);
+    // namespace + version, both 0 - 2 bytes
+    id.put_i16(0);
+
+    // host-based number - 5 bytes
+    lazy_static! {
+        static ref HOST_NUMBER: Vec<u8> = host_number();
+    }
+    id.put_slice(&HOST_NUMBER[..]);
+
+    // counter - 5 bytes
+    id.put_u8(0);
+    // one padding byte
+    // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
+    lazy_static! {
+        static ref COUNTER: CounterU32 = CounterU32::new(0);
+    }
+    id.put_u32(COUNTER.inc()); // 4 bytes
+
+    base64::encode(id.freeze())
+}
```

Comment:
This builds on my concerns above regarding `lazy_static` but I think illustrates a general theme: this function is too complicated because you need global state. Consider that if this function were embedded in a struct you wouldn't need to use an atomicu32 -- which is what `CounterU32` ends up being, except but with a mutex -- and you could elide the statics. 

The event encoding code here and related functions need to be refactored into a structure that encodes events, itself probably part of the sink state.

Reply:
Just to make sure - I moved all the state into a struct, but still using an AtomicU32 to enforce thread-safety, but I guess what you suggest is that using u32 would be enough, right? Sorry, my knowledge about concurrency in Rust is still a bit blurry.

Reply:
You don't need thread safety because there aren't multiple threads interacting with the same area of memory. For what it's worth you can always be sure that if you write code that isn't thread-safe Rust will fail to compile your program; it's one of the major safety guarantees. 

Maybe the question I have for you is, where do you there might be multiple threads interacting with this bit of memory? Hard to know how to explain this without addressing the more general misunderstanding, I think.

Reply:
yes, sorry, I just somehow expected that our sinks would be _not_ single-threaded, that's where all these thread-safety concerns are coming from, but now I know that it is not so we can use u32 safely.